mayastor/src/bdev/nexus/nexus_label.rs

```diff
@@ -126,113 +164,110 @@ impl Nexus {
                 .unwrap(),
             ent_guid: GptGuid::new_random(),
             ent_start: entries[0].ent_end + 1,
-            ent_end: hdr.lba_end,
+            ent_end: header.lba_end,
             ent_attr: 0,
             ent_name: GptName {
                 name: "MayaData".into(),
             },
         };
 
-        hdr.table_crc = GptEntry::checksum(&entries);
+        header.table_crc = GptEntry::checksum(&entries);
+
+        //
+        // Secondary GPT header
+        let mut backup = header.to_backup();
+
+        header.checksum();
+        backup.checksum();
 
         NexusLabel {
-            primary: hdr,
+            mbr: pmbr,
+            primary: header,
             partitions: entries,
+            secondary: backup,
         }
     }
 
-    /// write the protective MBR to all children.
-    pub async fn write_pmbr(&mut self) -> Result<(), LabelError> {
-        let mut pmbr = Pmbr::default();
-        let mut buf =
-            DmaBuf::new(self.bdev.block_len() as usize, self.bdev.alignment())
-                .context(WritePmbrAlloc {})?;
-
-        // the max size with MBR is 2GB, if we are smaller however, we must
-        // ensure that we reflect that in the MBR as well even though its not
-        // used.
-
-        let size = self.bdev.num_blocks() * self.bdev.block_len() as u64;
-        pmbr.entries[0].attributes = 0x00;
-        //
-        pmbr.entries[0].chs_start = [0x00, 0x02, 0x00];
-        pmbr.entries[0].chs_last = [0xff, 0xff, 0xff];
-
-        // this indicated that we are "protective MBR", saying we use all
-        // use all storage on this device.
-
-        pmbr.entries[0].ent_type = 0xee;
-        pmbr.entries[0].num_sectors = if size < u32::max_value().into() {
-            size as u32
-        } else {
-            u32::max_value()
-        };
-
-        pmbr.signature = [0x55, 0xaa];
-
+    fn get_primary_data(
+        &self,
+        label: &mut NexusLabel,
+    ) -> Result<LabelData, LabelError> {
+        let block_size = self.bdev.block_len() as u64;
+        let mut buf = DmaBuf::new(
+            (GPTHeader::PARTITION_TABLE_SIZE + 2 * block_size) as usize,
+            self.bdev.alignment(),
+        )
+        .context(WriteLabelAlloc {})?;
         let mut writer = Cursor::new(buf.as_mut_slice());
-        // we seek 440 into the buffer here, this makes serialisation a little
-        // easier.
 
+        // Protective MBR
         writer.seek(SeekFrom::Start(440)).unwrap();
-        serialize_into(&mut writer, &pmbr).context(SerializeError {})?;
-
-        for child in &mut self.children {
-            child.write_at(0, &buf).await.context(WriteError {})?;
+        serialize_into(&mut writer, &label.mbr).context(SerializeError {})?;
+
+        // Primary GPT header
+        writer
+            .seek(SeekFrom::Start(label.primary.lba_self * block_size))
+            .unwrap();
```

Comment:
should these unwraps be handled?

Reply:
This is how the code was originally so I did not change it. In this case we are writing (and seeking) to an in-memory buffer of known size, and so I guess it is not unreasonable to assume that they won't ever fail (if they do you have got bigger problems anyway).